gorp/gorp.go

```diff
@@ -0,0 +1,331 @@
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectInt(query string, args ...interface{}) (int64, error) {
+	var result int64
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectInt(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectNullInt(query string, args ...interface{}) (sql.NullInt64, error) {
+	var result sql.NullInt64
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectNullInt(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectFloat(query string, args ...interface{}) (float64, error) {
+	var result float64
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectFloat(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectNullFloat(query string, args ...interface{}) (sql.NullFloat64, error) {
+	var result sql.NullFloat64
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectNullFloat(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectStr(query string, args ...interface{}) (string, error) {
+	var result string
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectStr(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectNullStr(query string, args ...interface{}) (sql.NullString, error) {
+	var result sql.NullString
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.SelectNullStr(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		err = e.inner.SelectOne(holder, query, args...)
+		return err
+	})
+
+	return err
+}
+
+func (e *executor) Query(query string, args ...interface{}) (*sql.Rows, error) {
+	var result *sql.Rows
+	var err error
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result, err = e.inner.Query(query, args...)
+		return err
+	})
+
+	return result, err
+}
+
+func (e *executor) QueryRow(query string, args ...interface{}) *sql.Row {
+	var result *sql.Row
+
+	jgorp.Trace(e.ctx, e.name, format(query, args), func() error {
+		result = e.inner.QueryRow(query, args...)
+		return nil
+	})
+
+	return result
+}
+
+func format(query string, args []interface{}) string {
+	re := regexp.MustCompile("\\$(\\d+)")
+	template := re.ReplaceAllString(query, "%[$1]v")
+	return fmt.Sprintf(template, args...)
+}
+
+func types(list []interface{}) []reflect.Type {
+	var result []reflect.Type
+	for _, val := range list {
+		result = append(result, reflect.TypeOf(val))
```

Comment:
Using reflect is usually an anti-patter in go, since it's quite expensive. I think this won't be the bottleneck in our db calls (the queries are quite big) so I'm ok with it. We just need to keep track if this is not impacting the application.

Reply:
Would a `fmt.Sprintf("%T", val)` be less expensive, or does it use reflect internally?

Reply:
It uses reflect internally: https://github.com/golang/go/blob/master/src/fmt/print.go#L626